tools/lox_gen/src/main.rs

```diff
@@ -1,69 +1,298 @@
+use std::fs;
 use std::path::Path;
-use std::{fs, io, io::prelude::*};
-
-type Error = Box<dyn std::error::Error + Send + Sync>;
-type Result<T> = std::result::Result<T, Error>;
-
-pub fn main() -> Result<()> {
-    let mut output = vec!["// AUTO-GENERATED DO NOT EDIT!".to_string(), "".to_string()];
-    let mut in_data = false;
-    let mut statement = Vec::new();
-    for line in io::stdin().lock().lines() {
-        let l = line?;
-        if l.is_empty() {
-            continue;
+use std::process::Command;
+
+use proc_macro2::{Ident, TokenStream};
+use quote::{format_ident, quote};
+
+use lox_io::spice::Kernel;
+use naif_ids::{Body, BARYCENTERS, MINOR_BODIES, PLANETS, SATELLITES, SUN};
+
+mod naif_ids;
+
+type Generator = fn(
+    imports: &mut Vec<Ident>,
+    code: &mut TokenStream,
+    tests: &mut TokenStream,
+    ident: &Ident,
+    id: &i32,
+    data: &Data,
+);
+
+struct Data {
+    pck: Kernel,
+    gm: Kernel,
+}
+
+pub fn main() {
+    let pck = Kernel::from_string(include_str!("../../../data/pck00011.tpc"))
+        .expect("parsing should succeed");
+    let gm = Kernel::from_string(include_str!("../../../data/gm_de440.tpc"))
+        .expect("parsing should succeed");
+    let data = Data { pck, gm };
+    let bodies: [(&str, Vec<Body>, Vec<Generator>); 5] = [
+        ("sun", Vec::from(SUN), vec![naif_id, point_mass, spheroid]),
+        (
+            "barycenters",
+            Vec::from(BARYCENTERS),
+            vec![naif_id, point_mass],
+        ),
+        (
+            "planets",
+            Vec::from(PLANETS),
+            vec![naif_id, point_mass, spheroid],
+        ),
+        (
+            "satellites",
+            Vec::from(SATELLITES),
+            vec![naif_id, point_mass, tri_axial],
+        ),
+        (
+            "minor",
+            Vec::from(MINOR_BODIES),
+            vec![naif_id, point_mass, tri_axial],
+        ),
+    ];
+    bodies
+        .iter()
+        .for_each(|(file, bodies, generators)| write_file(file, bodies, generators, &data));
+}
+
+fn write_file(file: &str, bodies: &[Body], generators: &[Generator], data: &Data) {
+    let path = format!("../../../crates/lox_core/src/bodies/{file}.rs");
+    let mut code = String::from("// Auto-generated by `lox_gen`. Do not edit!\n");
```

Comment:
Can be moved to the `quote!` in line 90.

Reply:
This was the only way I found to add a comment to the file (which is not a doc comment). A comment in the `quote!` is still a comment :grimacing: and I did not see an escape hatch in the docs.

Reply:
Lol. Perhaps in the future you can just add a comment to explain why you took this approach and not the other